### DISCUSSION TITLE
RFC: Tape Gateway

### DISCUSSION THREAD

**Discussion by gpmidi** [minio/minio#14771] [Q&A] [answered]
# Overview
I'm pondering the idea of writing a tape [gateway](https://github.com/minio/minio/tree/master/cmd/gateway) - My thinking is to have the gateway code operate as a organizer for a like a set of NAS-like or "normal-like" Minio server instances. Storage wise there would be one per tape accessed via a fully cached LTFS mount. 

# Reads
The read flow would look something like this: A request would hit the gateway server and it'd check for it in the disk backed caches.  Assuming that object version isn't found in the cache it would then check a database (details TBD) for what site(s)-library(s)-tape(s) have copies of the object-versions. It'd the use an API call to a daemon that would load the given tape to a free drive, mount it via LTFS with a full preread cache (mount flag), and then kick off a MinIO instance to use that mount. 

# Assumptions
1. I'll be doing a fair bit of coding
2. Write heavy loads - Only for archival and backups - The cache making up for any metadata type objects. 
3. There is a strong preference to have this be strong enough to be merged in upstream or at least structured in such a way that maintaining a fork isn't horrifically painful. 
4. There will be a lot of tuning timeouts and such to give the library time to load a tape and a system to read the full tape into an SSD based cache before the Minio instance can start reading. 

# Questions 
1. Is there a reasonable way to cache all writes and only flush when it gets over some number of TiB greater than a tape? 
2. Is there a reasonable way for the Gateway code to proactively move other objects (ie everything on the tape being read) into the cache? The idea here being that whenever a tape is loaded all of the objects on the tape would get cached since the odds some of them might be needed soon after is high. 
3. How dumb? Thoughts? Ideas?

**Comment by harshavardhana** [comment, marked as answer]
Gateway itself is being dropped #14331 there is no future in gateway mode. Please move to MinIO server deployments and then tier to Cloud.

**Reply by gpmidi** [reply]
So base it around the now named 'edge' code instead?

**Reply by harshavardhana** [reply]
You can write your own and keep it in your fork, we are going to remove gateway code.

**Reply by gpmidi** [reply]
> S3 Gateway is becoming minio edge for edge deployments with disk caching

I could modify/abuse that code

Thoughts?

**Reply by gpmidi** [reply]
@harshavardhana To be fair here, I'm trying to figure out a solution to make tape via S3-like APIs viable. I've got alternatives playing with FUSE or (eww) RADIOS. But having it in/near Minio seems to make a lot more sense. 
Tape is so much cheaper per TiB-Month than any other solution out there for backups and archival. I couldn't afford over a PiB of backup space any other way.

**Reply by gpmidi** [reply]
@harshavardhana The to-be-removed gateway code is [here](https://github.com/minio/minio/tree/master/cmd/gateway), right? Or is there some other repo I missed?

Also - A glacial like API sounds nice but wouldn't meet most people's needs - S3 is more sensible :(

**Reply by harshavardhana** [reply]
> > S3 Gateway is becoming minio edge for edge deployments with disk caching

You can but the use-case there is limited and we are not planning for further expanding scope here. Its not going to get any divergent features like TAPE.

**Reply by gpmidi** [reply]
Guess it comes down to maintaining a fork with that in it or maybe writing something new with Minio as a library. I'll hack at it some and see where I get. 

Thanks for the input.

**Reply by gpmidi** [reply]
Side note - Now that I've finally finished moving and got my three Scalar i6k libraries back up and running - I can FINALLY get to work on this in a serious way :-D

**Reply by D4rk4** [reply]
@gpmidi any progress with LTFS?

**Reply by gpmidi** [reply]
> @gpmidi any progress with LTFS?

I've ended up going a totally different route after the less than warm response here - Using a rclone frontend and an API plus a TBD system for moving the file data itself. 

The TBD system might be S3 like or might be something else. Doesn't really matter, just needs to give the server side and the rclone clients a way to read/write without going through the api itself. Although that would technically be an option too if I wanted.
